### PR TITLE
Feat/ertb keyword tracking

### PR DESCRIPTION
**Pull request by t-eliks**
Set up keyword tracking for ExtendedRichTextBox control. Allows / commands, ex. /todo.

Prerequisite for insertion of TODOs without a button

### CODE REVIEW

**Comment by t-eliks** [comment]
[Add base TODO items](https://trello.com/c/FmMgIkNN/11-add-base-todo-items)

**Comment by t-eliks** [comment]
[Allow inserting TODOs anywhere](https://trello.com/c/cGs43i1W/13-allow-inserting-todos-anywhere)

**Comment by t-eliks** [comment]
[Add an ability to insert a TODO item without clicking the button](https://trello.com/c/F1GddRqj/28-add-an-ability-to-insert-a-todo-item-without-clicking-the-button)

File: ToNote/Controls/ExtendedRichTextBox.cs [hunk 1]
```diff
@@ -9,21 +11,78 @@ public class ExtendedRichTextBox : RichTextBox
...
-                    var range = new TextRange(Document.ContentStart, Document.ContentEnd);
                     if (string.IsNullOrWhiteSpace(range.Text))
                         BackspacePressedWhileEmpty?.Invoke(this, new RoutedEventArgs());
                 }
             };
 
+            this.TextChanged += (s, e) =>
+            {
+                //Prevents start of tracking while initializing.
+                if (this.IsLoaded)
```
Comment (Kodnot) [inline ToNote/Controls/ExtendedRichTextBox.cs]: Invert if to reduce nesting

File: ToNote/Controls/ExtendedRichTextBox.cs [hunk 2]
```diff
@@ -9,21 +11,78 @@ public class ExtendedRichTextBox : RichTextBox
...
+                        {
+                            _tracking = false;
+                            return;
+                        }
+
+                        var keyword = range.Text.Substring(_slashIndex + 1, offset - _slashIndex);
+
+                        foreach (var keywordaction in _trackedKeywords)
+                        {
+                            if (keyword == keywordaction.Keyword)
```
Comment (Kodnot) [inline ToNote/Controls/ExtendedRichTextBox.cs]: camelCase, `keywordaction` -> `keywordAction`

File: ToNote/Controls/ExtendedRichTextBox.cs [hunk 3]
```diff
@@ -9,21 +11,78 @@ public class ExtendedRichTextBox : RichTextBox
...
+                        {
+                            _tracking = false;
+                            return;
+                        }
+
+                        var keyword = range.Text.Substring(_slashIndex + 1, offset - _slashIndex);
+
+                        foreach (var keywordaction in _trackedKeywords)
+                        {
+                            if (keyword != keywordaction.Keyword + ' ') return;
```
Comment (Kodnot) [inline ToNote/Controls/ExtendedRichTextBox.cs]: Also add newline as a possible suffix of the keyword. Typing "/note" + enter feels natural, but currently does not work.

File: ToNote/Controls/ExtendedRichTextBox.cs [hunk 4]
```diff
@@ -9,21 +11,78 @@ public class ExtendedRichTextBox : RichTextBox
...
+                            //Trims the '/' + keyword from the end of the text
+                            range.Text = range.Text.Remove(offset - keyword.Length, keyword.Length + 1);
+
+                            keywordaction.Action.Invoke();
+
+                            return;
+                        }
+                    }
+
+                    if (lastChar == '/' && secondToLastChar == ' ')
```
Comment (Kodnot) [inline ToNote/Controls/ExtendedRichTextBox.cs]: Should also work when the second to last char is new line (when the keyword is entered at the start of a new line). Currently this works only sometimes. If you create a new RTB and enter "/note ", the keyword will trigger properly. However, if you create a new RTB, press enter and then type "/note ", the keyword will not trigger.

File: ToNote/Controls/ExtendedRichTextBox.cs [hunk 5]
```diff
@@ -39,5 +98,18 @@ public void ReadFromFile(string file)
 
             CurrentFile = file;
         }
+
+        public void TrackKeyword(string keyword, Action action)
+        {
+            var keywordaction = new KeywordAction() { Keyword = keyword, Action = action };
```
Comment (Kodnot) [inline ToNote/Controls/ExtendedRichTextBox.cs]: [Nitpick] `keywordaction` -> `keywordAction`

File: ToNote/Controls/ExtendedRichTextBox.cs [hunk 6]
```diff
@@ -39,5 +98,18 @@ public void ReadFromFile(string file)
...
+            if (!_trackedKeywords.Contains(keywordaction))
```
Comment (Kodnot) [inline ToNote/Controls/ExtendedRichTextBox.cs]: I think this check is redundant. `KeywordAction` contains an `Action`. For a comparison between two `KeywordAction` instances to return true, they would have to reference the same action (default reference equality comparison). One would have to go out of their way to trigger this check, for example even just copy-paste duplicating the "note" keyword registering code does not trigger this condition.
If you think that one shoud only be able to register one action per keyword, overload the equality comparison operators for `KeywordAction`. 
However, I don't really see why you shouldn't be able to register multiple actions per keyword (unless maybe there could be some issues with execution order?). If we allow this, you can just remove this check, but then also don't forget to remove `return` from the foreach where you search for maching keywords.

**Review by Kodnot — CHANGES_REQUESTED** [review]

File: ToNote/Controls/ExtendedRichTextBox.cs [hunk 7]
```diff
@@ -9,21 +11,77 @@ public class ExtendedRichTextBox : RichTextBox
...
+                        if (keyword != keywordAction.Keyword + ' ' && keyword != keywordAction.Keyword + '\r' + '\n') continue;
+
+                        _tracking = false;
+
+                        //Trims the '/' + keyword from the end of the text
+                        range.Text = range.Text.Remove(offset - keyword.Length, keyword.Length + 1);
+
+                        keywordAction.Action.Invoke();
+
+                        return;
```
Comment (Kodnot) [inline ToNote/Controls/ExtendedRichTextBox.cs]: Remove this return, as you currently allow registering multiple keywords per action, but only execute the first found action.

**Review by Kodnot — APPROVED** [review]

**Pull request merged**